src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseStream.cs

```diff
@@ -164,10 +166,20 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
                 }, 
                 CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
 
-            // TODO: Issue #2165. Register callback on cancellation token to cancel WinHTTP operation.
-                
+            // Register callback on cancellation token to cancel any pending WinHTTP operation.
+            if (token != CancellationToken.None)
+            {
+                WinHttpTraceHelper.Trace("WinHttpResponseStream.ReadAsync: registering for cancellation token request");
+                _ctr = token.Register(() => CancelPendingResponseStreamReadOperation());
```

Comment:
token.Register(s => ((WinHttpResponseStream)s).CancelPendingResponseStreamReadOperation(), this)